### ISSUE TITLE
Safely deploying new Compute VMs for December 2018 DSG

### ISSUE THREAD

**Issue by martintoreilly** [alan-turing-institute/data-safe-haven#138]
The issue is to implement the change control process described in #87 

## Base image for Dec 2018 DSG
- The base image is version `0.0.2018120701` of the `Ubuntu` Gallery image. The initial (Sunday evening). Shared VM deployments were made directly from this image and any custom images will be based on this image.
-  See below for change control process plus instructions and scripts for safely deploying additional VMs to DSG groups.

### Testing / staging environments
- DSGTest (testing):`10.250.250.160` (VM `DSG201812091917` in RG `RG_DSG_COMPUTE`)
- DSG6 (validation):`` (VM `` in RG `RG_DSG_COMPUTE`)

### Production environments
- DSG1: `10.250.2.160` (VM `DSG201812091935` in RG `RG_DSG_COMPUTE`)
- DSG2: `10.250.10.160` (VM `DSG201812091941` in RG `RG_DSG_COMPUTE`)
- DSG3: `10.250.18.160` (VM `DSG201812091945` in RG `RG_DSG_COMPUTE`)
- DSG4: `10.250.26.160` (VM `DSG201812091930` in RG `RG_DSG_COMPUTE`)

### DSG Team
- Create a new issue with the requested package(s). There should be a new issue for each set of requested packages for **each DSG group** that needs them. This will ensure we test the exact VMs that will be deployed for each DSG group even if the cumulative list of approved additional packages diverges across groups. The non-technical decision process on whether to add the package(s) happens in each issue and may be different for different DSG groups due to differences in the impact of not having these packages. Only if the outcome of this process is to approve the package installation will we proceed to the next step.
### REG team
- Create a new branch from the `DSG-DEC2018` branch, named `Dec18-DSG-X-Issue-Y`, where `X` is the DSG group number and `Y` is the issue number.
- Update the `cloud-init` build file for that **specific** DSG group to add the approved packages. 
- Commit the updated `cloud-init` build file for that **specific** DSG group 
- Open a PR to the `DSG-DEC2018` branch and get a second member of the team to review the changes.
- Ensure the issue branch is up to date with the `DSG-DEC2018`
- Test the updates by building a VM using the updated `cloud-init` build file in the DSG testing environment.
- Get an appropriate member of the DSG group to test the VM with the new packages in the test environment (on your laptop - do **not** add them to the test group)
- Record the successful testing in the PR comments
- Register the successfully tested VM image as a new Gallery image for that **specific** DSG group.
- Update the DSG deploy script to use the newly registered Gallery image version for that **specific** DSG group.
- Ensure the issue branch is up to date with the `DSG-DEC2018` branch, pulling changes and re-running testing if not.
- Merge the PR into the `DSG-DEC2018` branch
- Deploy a new VM using the DSG deploy script from the `DSG-DEC2018` branch

### Copying common data
- Create a shared folder on the VM
- SSH into the VM with the local admin account
- Create a shared data folder with `sudo mkdir /data`
- Temporarily give all users read and write access with `sudo chmod go+rw /data`
- Run the WinSCP "File transfer" app as your normal AD user
- Connect to the VM's IP address
- Copy the data files in the "R" drive to the "/data" folder on the VM

### Notes
The above process has supports a **single** custom VM configuration for each DSG group at any point in time. All new VMs deployed will use the **latest** VM configuration for that **specific** DSG group.

## Deploying a new VM
Deploy to DSG1-4, DSG6 and DSGTest using the `deploy_dsg_dec18.sh` script.

**Warning:** You cannot deploy a VM to more than one DSG environment in parallel. This is because the subscription used by the `az` command line tool is set globally.

### DSG deploy script usage
```
usage: ./deploy_dsg_dec18.sh -g dsg_group_id [-h] [-i source_image] [-x source_image_version] [-z vm_size]
  -h                        display help
  -d dsg_group_id           specify the DSG group to deploy to ('TEST' for test or 1-6 for production)
  -z vm_size                specify a VM size to use (defaults to 'Standard_DS2_v2')
  -q fixed_ip               Last part of IP address (first three parts are fixed for each DSG group)

```

### Examples
- To deploy a default sized (small) VM to DSG6 with a dynamic IP for testing: `./deploy_dsg_dec18.sh -d 6`
- To deploy a small sized VM to DSG6 ay a fixed IP of `10.250.250.X: `./deploy_dsg_dec18.sh -d 6 -z Standard_D2_v3 -q X`

#### Dsv3-series 1 (General compute - CPU only)
Limit is 350 vCPUs total per DSG
ACU: 160-190 | Premium Storage: Supported | Premium Storage Caching: Supported

Dsv3-series sizes are based on the 2.4 GHz Intel Xeon® E5-2673 v3 (Haswell) processor or the latest 2.3 GHz Intel XEON ® E5-2673 v4 (Broadwell) processor that can achieve 3.5GHz with Intel Turbo Boost Technology 2.0 and use premium storage. The Dsv3-series sizes offer a combination of vCPU, memory, and temporary storage for most production workloads.

Size | vCPUs | Memory: GiB | Temp storage (SSD) GiB | Max data disks | Max cached and temp storage throughput: IOPS / MBps (cache size in GiB) | Max uncached disk throughput: IOPS / MBps | Max NICs / Expected network bandwidth (Mbps)
-- | -- | -- | -- | -- | -- | -- | --
Standard_D2s_v3 | 2 | 8 | 16 | 4 | 4,000 / 32 (50) | 3,200 / 48 | 2 / 1,000
Standard_D4s_v3 | 4 | 16 | 32 | 8 | 8,000 / 64 (100) | 6,400 / 96 | 2 / 2,000
Standard_D8s_v3 | 8 | 32 | 64 | 16 | 16,000 / 128 (200) | 12,800 / 192 | 4 / 4,000
Standard_D16s_v3 | 16 | 64 | 128 | 32 | 32,000 / 256 (400) | 25,600 / 384 | 8 / 8,000
Standard_D32s_v3 | 32 | 128 | 256 | 32 | 64,000 / 512 (800) | 51,200 / 768 | 8 / 16,000
Standard_D64s_v3 | 64 | 256 | 512 | 32 | 128,000 / 1024 (1600) | 80,000 / 1200

#### Fsv2-series 1 (Compute optimised - CPU only)
Limit is 350 vCPUs total per DSG
ACU: 195 - 210 | Premium Storage: Supported | Premium Storage Caching: Supported

Size | vCPU's | Memory: GiB | Temp storage (SSD) GiB | Max data disks | Max cached and temp storage throughput: IOPS / MBps (cache size in GiB) | Max uncached disk throughput: IOPS / MBps | Max NICs / Expected network bandwidth (Mbps)
-- | -- | -- | -- | -- | -- | -- | --
Standard_F2s_v2 | 2 | 4 | 16 | 4 | 4000 / 31 (32) | 3200 / 47 | 2 / 875
Standard_F4s_v2 | 4 | 8 | 32 | 8 | 8000 / 63 (64) | 6400 / 95 | 2 / 1,750
Standard_F8s_v2 | 8 | 16 | 64 | 16 | 16000 / 127 (128) | 12800 / 190 | 4 / 3,500
Standard_F16s_v2 | 16 | 32 | 128 | 32 | 32000 / 255 (256) | 25600 / 380 | 4 / 7,000
Standard_F32s_v2 | 32 | 64 | 256 | 32 | 64000 / 512 (512) | 51200 / 750 | 8 / 14,000
Standard_F64s_v2 | 64 | 128 | 512 | 32 | 128000 / 1024 (1024) | 80000 / 1100 | 8 / 28,000

#### NC-series (Tesla GPUs)
Limit is 48 vCPUs total per DSG (i.e. 8 GPUs)
Premium Storage: Not Supported | Premium Storage Caching: Not Supported

NC-series VMs are powered by the NVIDIA Tesla K80 card. Users can crunch through data faster by leveraging CUDA for energy exploration applications, crash simulations, ray traced rendering, deep learning, and more. The NC24r configuration provides a low latency, high-throughput network interface optimized for tightly coupled parallel computing workloads.

Size | vCPU | Memory: GiB | Temp storage (SSD) GiB | GPU | GPU memory: GiB | Max data disks | Max NICs
-- | -- | -- | -- | -- | -- | -- | --
Standard_NC6 | 6 | 56 | 340 | 1 | 8 | 24 | 1
Standard_NC12 | 12 | 112 | 680 | 2 | 16 | 48 | 2
Standard_NC24 | 24 | 224 | 1440 | 4 | 32 | 64 | 4
Standard_NC24r* | 24 | 224 | 1440 | 4 | 32 | 64 | 4
